torch/distributed/distributed_c10d.py

```diff
@@ -753,10 +755,35 @@ def recv(tensor,
         return src
 
 
+def async_collective(func):
```

Comment:
As @rohan-varma mentioned in a separate thread, we could call `process_group->allreduce()` from C++ (ex: c10d/reducer.cpp). Let's put all the RecordFunction logic in ProcessGroupGloo.cpp and ProcessGroupNCCL.cpp to ensure pure C++ invocations are profiled as well.